pjrt/convert/trace_container.zig

```diff
@@ -0,0 +1,239 @@
+                    try event.args.put(allocator, xplane.getStatMetadata(xstat.metadata_id).name.getSlice(), stat_str);
+                }
+            }
+        }
+    }
+
+    fn getTraceViewerMaxEvents() !u64 {
+        const kMaxEvents = 1000000;
+        if (std.posix.getenv("TF_PROFILER_TRACE_VIEWER_MAX_EVENTS")) |max_events| {
+            return std.fmt.parseInt(u64, max_events, 10);
+        } else return kMaxEvents;
+    }
+
+    pub fn fromXSpace(allocator: std.mem.Allocator, xspace: *const xplane_proto.XSpace) !TraceContainer {
+        var self: TraceContainer = .{};
+        if (findPlaneWithName(xspace, host_threads_plane_name)) |hp| {
+            const xplane = try xplane_visitor.XPlaneVisitor.init(
+                allocator,
+                hp,
+                xplane_schema.HostEventType.fromString,
+                xplane_schema.StatType.fromString,
+            );
+            try self.xplaneToTraceEvents(allocator, host_threads_device_id, &xplane);
+        }
+
+        var device_planes = try findPlanesWithPrefix(allocator, xspace, gpu_plane_prefix);
+
+        // We don't expect GPU and TPU planes and custom devices to be present in the
+        // same XSpace.
+        if (device_planes.len == 0) {
+            device_planes = try findPlanesWithPrefix(allocator, xspace, tpu_plane_prefix);
+        }
+        if (device_planes.len == 0) {
+            device_planes = try findPlanesWithPrefix(allocator, xspace, custom_plane_prefix);
+        }
+
+        for (device_planes) |dp| {
+            const xplane = try xplane_visitor.XPlaneVisitor.init(
+                allocator,
+                dp,
+                xplane_schema.HostEventType.fromString,
+                xplane_schema.StatType.fromString,
+            );
+            const device_id: u32 = first_device_id + @as(u32, @intCast(xplane.plane.id));
+            try self.xplaneToTraceEvents(allocator, device_id, &xplane);
+        }
+
+        // Trace viewer (non-streaming) has scalability issues, we need to drop
+        // events to avoid loading failure for trace viewer.
+        const viewer_max_events = try getTraceViewerMaxEvents();
+        self.capEvents(viewer_max_events);
+        return self;
+    }
+
+    pub fn createEvent(self: *TraceContainer, allocator: std.mem.Allocator) !*TraceEvent {
+        const event = try allocator.create(TraceEvent);
+        event.* = .{};
+        try self.events.append(allocator, event);
+        return event;
+    }
+
+    pub fn capEvents(self: *TraceContainer, max_count: u64) void {
+        const total_count = self.events.items.len;
+        if (total_count <= max_count) {
+            // Nothing to do. Events are not known sorted after return.
+            return;
+        }
+        // sort the events according to start time.
+        // TODO: partial sort would improve performance.
+        std.mem.sort(*TraceEvent, self.events.items, {}, struct {
```

Comment:
it feels a bit late to start dropping events. Maybe we should instead put a limit earlier on while reading events from the protobuffer. But for now it's ok to assume we can handle the input trace file. Ultimately the user control the size of the trace based on how long they run the model.